roboteam_ai/src/coach/midField/HarassRobotCoach.cpp

```diff
@@ -0,0 +1,270 @@
+        }
+    }
+    // else, aim towards the ball
+    return (ball->pos - thisRobot->pos).toAngle();
+}
+
+HarassRobotCoach::HarassTarget HarassRobotCoach::findRobotToHarass(const RobotPtr &thisRobot, int &myIndex,
+        bool goAfterBall) {
+    HarassTarget harassTarget;
+
+    // check opponent closest to ball and closest to harasser
+    auto closestRobotToBallDistance = DBL_MAX;
+    RobotPtr closestRobotToBall;
+
+    auto closestRobotToHarasserDistance = DBL_MAX;
+    RobotPtr closestRobotToHarasser;
+
+    for (auto robot : world::world->getThem()) {
+        if (abs(robot.pos.x) <= HARASS_THRESHOLD) {
+            if (goAfterBall) {
+                // check if robot is closest to ball
+                double distanceToBall = (ball->pos - robot.pos).length();
+                if (distanceToBall < closestRobotToBallDistance) {
+                    closestRobotToBallDistance = distanceToBall;
+                    closestRobotToBall = std::make_shared<Robot>(robot);
+                }
+            }
+
+            // check if robot is closest to the harasser
+            double distanceToHarasser = (thisRobot->pos - robot.pos).length();
+            if (distanceToHarasser < closestRobotToHarasserDistance) {
+                closestRobotToHarasserDistance = distanceToHarasser;
+                closestRobotToHarasser = std::make_shared<Robot>(robot);
+            }
+        }
+    }
+
+    if (goAfterBall) {
+        // see if opponent is too close to the ball
+        if (closestRobotToBallDistance <= TOO_CLOSE_TO_BALL_DISTANCE) {
+            // if this robot is closest to the robot with ball, harass that robot
+            int bestIndex = getRobotIndexCloseToEnemyRobot(closestRobotToBall);
+            if (bestIndex == myIndex) {
+                if (! robotAlreadyBeingHarassed(myIndex, closestRobotToBall->id)) {
+                    targetRobotsToHarass[myIndex] = closestRobotToBall;
+                    harassTarget.harassRobot = closestRobotToBall->id;
+                    harassTarget.harassPosition = harassRobot(myIndex, closestRobotToBall->id);
+                    return harassTarget;
+                }
+            }
+        }
+    }
+
+    // else, harass the opponent closest to the harasser
+
+
+    if(!robotAlreadyBeingHarassed(myIndex, closestRobotToHarasser->id)) {
+        targetRobotsToHarass[myIndex] = closestRobotToHarasser;
+        harassTarget.harassRobot = closestRobotToHarasser->id;
+        harassTarget.harassPosition = harassRobot(myIndex, closestRobotToHarasser->id);
+        return harassTarget;
+    }
+
+    // can't get a robot to harass, remove my targetRobot if it was there
+    targetRobotsToHarass[myIndex] = RobotPtr(nullptr);
+
+    harassTarget.harassRobot = -1;
+    harassTarget.harassPosition = standInMidField(thisRobot, myIndex);
+    return harassTarget;
+}
+
+Vector2 HarassRobotCoach::standInMidField(const HarassRobotCoach::RobotPtr &thisRobot, int &myIndex) {
```

Comment:
this function needs to be split in multiple ones.

Reply:
done